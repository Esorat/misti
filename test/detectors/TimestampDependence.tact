contract TimestampTest {
    value: Int;
    init() {
        self.value = now();  //BAD in Assignment
    }

    // Test basic variable declaration
    receive("test1") {
        let time: Int = now();  //BAD in var declaration
        self.value = time;
    }

    receive("foobar") {
        let v: Int = 0;
        if (self.value == 3) { // Bad
            v = now();
        }
        if (v > 3) { // Bad
            v += 1;
        }
        let v1: Int = v;
        if (v1 > 500) {} // Bad
        let v2: Int = 42;
        if (true) { v2 = v1; }
        if (v2 > 5) {} // Bad
    }

    // Test loops
    receive("test2") {
        let start: Int = now();  //BAD in var declaration
        
        // While loop
        while (now() < start + 1000) {
            self.value += 1;
        }
        
        // Repeat loop
        repeat(10) {
            self.value = now();
        }
        
        // Until loop
        do {
            self.value = self.value + 1;
        } until (now() > start + 2000);
    }

    // Test returns
    fun getTimestamp(): Int {
        return now();
    }

    // Test augmented assignment
    receive("test3") {
        self.value += now();
    }

    // Test expressions
    receive("test4") {
        if (self.value != 0) {
            self.value = now() * 2;
        }
    }
}