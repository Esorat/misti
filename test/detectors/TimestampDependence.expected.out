[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:4:9:
  3 |     init() {
> 4 |         self.value = now();  //BAD in Assignment
              ^
  5 |     }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:9:9:
   8 |     receive("test1") {
>  9 |         let time: Int = now();  //BAD in var declaration
               ^
  10 |         self.value = time;
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:10:9:
   9 |         let time: Int = now();  //BAD in var declaration
> 10 |         self.value = time;
               ^
  11 |     }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Tainted timestamp used in a condition
test/detectors/TimestampDependence.tact:15:9:
  14 |         let v: Int = 0;
> 15 |         if (self.value == 3) { // Bad
               ^
  16 |             v = now();
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:16:13:
  15 |         if (self.value == 3) { // Bad
> 16 |             v = now();
                   ^
  17 |         }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Tainted timestamp used in a condition
test/detectors/TimestampDependence.tact:18:9:
  17 |         }
> 18 |         if (v > 3) { // Bad
               ^
  19 |             v += 1;
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:19:13:
  18 |         if (v > 3) { // Bad
> 19 |             v += 1;
                   ^
  20 |         }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:21:9:
  20 |         }
> 21 |         let v1: Int = v;
               ^
  22 |         if (v1 > 500) {} // Bad
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Tainted timestamp used in a condition
test/detectors/TimestampDependence.tact:22:9:
  21 |         let v1: Int = v;
> 22 |         if (v1 > 500) {} // Bad
               ^
  23 |         let v2: Int = 42;
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Tainted timestamp used in a condition
test/detectors/TimestampDependence.tact:24:9:
  23 |         let v2: Int = 42;
> 24 |         if (true) { v2 = v1; }
               ^
  25 |         if (v2 > 5) {} // Bad
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:24:21:
  23 |         let v2: Int = 42;
> 24 |         if (true) { v2 = v1; }
                           ^
  25 |         if (v2 > 5) {} // Bad
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Tainted timestamp used in a condition
test/detectors/TimestampDependence.tact:25:9:
  24 |         if (true) { v2 = v1; }
> 25 |         if (v2 > 5) {} // Bad
               ^
  26 |     }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:30:9:
  29 |     receive("test2") {
> 30 |         let start: Int = now();  //BAD in var declaration
               ^
  31 |         
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Loop condition depends on tainted timestamp
test/detectors/TimestampDependence.tact:33:9:
  32 |         // While loop
> 33 |         while (now() < start + 1000) {
               ^
  34 |             self.value += 1;
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Loop condition depends on tainted timestamp
test/detectors/TimestampDependence.tact:38:9:
  37 |         // Repeat loop
> 38 |         repeat(10) {
               ^
  39 |             self.value = now();
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:39:13:
  38 |         repeat(10) {
> 39 |             self.value = now();
                   ^
  40 |         }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Loop condition depends on tainted timestamp
test/detectors/TimestampDependence.tact:43:9:
  42 |         // Until loop
> 43 |         do {
               ^
  44 |             self.value = self.value + 1;
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Returning a tainted timestamp
test/detectors/TimestampDependence.tact:50:9:
  49 |     fun getTimestamp(): Int {
> 50 |         return now();
               ^
  51 |     }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:55:9:
  54 |     receive("test3") {
> 55 |         self.value += now();
               ^
  56 |     }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Tainted timestamp used in a condition
test/detectors/TimestampDependence.tact:60:9:
  59 |     receive("test4") {
> 60 |         if (self.value != 0) {
               ^
  61 |             self.value = now() * 2;
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:61:13:
  60 |         if (self.value != 0) {
> 61 |             self.value = now() * 2;
                   ^
  62 |         }
Conditional logic depends on `now()`.
Help: Detailed check variable now() base on your code. Now() is predictable. Block timestamps are publicly visible, making them unsuitable for generating random numbers.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence